Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs

```diff
@@ -0,0 +1,744 @@
+        /// Only use if you know that your given pointer id corresponds to a pointer that has grabbed
+        /// this component.
+        /// </summary>
+        public Vector3 GetPointerGrabPoint(uint pointerId)
+        {
+            Assert.IsTrue(pointerIdToPointerMap.ContainsKey(pointerId));
+            return pointerIdToPointerMap[pointerId].GrabPoint;
+        }
+
+        #endregion Public Methods
+
+        #region Hand Event Handlers
+
+        /// <inheritdoc />
+        public void OnPointerDown(MixedRealityPointerEventData eventData)
+        {
+            if (!allowFarManipulation && eventData.Pointer as IMixedRealityNearPointer == null)
+            {
+                return;
+            }
+
+            // If we only allow one handed manipulations, check there is no hand interacting yet. 
+            if (manipulationType != HandMovementType.OneHanded || pointerIdToPointerMap.Count == 0)
+            {
+                uint id = eventData.Pointer.PointerId;
+                // Ignore poke pointer events
+                if (!eventData.used && !pointerIdToPointerMap.ContainsKey(id))
+                {
+                    // cache start ptr grab point
+                    pointerIdToPointerMap.Add(id, new PointerData(eventData.Pointer, eventData.Pointer.Result.Details.Point));
+
+                    // Call manipulation started handlers
+                    var handsPressedCount = pointerIdToPointerMap.Count;
+                    if (manipulationType.HasFlag(HandMovementType.TwoHanded) && handsPressedCount > 1)
+                    {
+                        if (!manipulationType.HasFlag(HandMovementType.OneHanded))
+                        {
+                            HandleManipulationStarted();
+                        }
+                        HandleTwoHandManipulationStarted();
+                    }
+                    else if (manipulationType.HasFlag(HandMovementType.OneHanded) && handsPressedCount == 1)
+                    {
+                        if (!isManipulationStarted)
+                        {
+                            HandleManipulationStarted();
+                        }
+                        HandleOneHandMoveStarted();
+                    }
+                }
+            }
+
+            if (pointerIdToPointerMap.Count > 0)
+            {
+                // Always mark the pointer data as used to prevent any other behavior to handle pointer events
+                // as long as the ManipulationHandler is active.
+                // This is due to us reacting to both "Select" and "Grip" events.
+                eventData.Use();
+            }
+        }
+
+        public void OnPointerDragged(MixedRealityPointerEventData eventData)
+        {                    
+            // Call manipulation updated handlers
+            var handsPressedCount = pointerIdToPointerMap.Count;
+            if (manipulationType.HasFlag(HandMovementType.OneHanded) && handsPressedCount == 1)
```

Comment:
>manipulationType.HasFlag(HandMovementType.OneHanded) && handsPressedCount == 1 [](start = 16, length = 78)

There's a decent amount of copy/paste instances of this + the two hand case (i.e. twohanded + > 1) - consider having a helper function like IsOneHandedManipulationActive() or IsTwoHandedManipulationActive() to avoid the copy paste.

Reply:
Active actually may be a bad word because it's also used in the pervious function to also start the manipulation as well, but yeahhhh

---
In reply to: [324915287](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5955#discussion_r324915287) [](ancestors = 324915287)

Reply:
Or maybe not, maybe this isn't used too much down below (still reading)

---
In reply to: [324915398](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5955#discussion_r324915398) [](ancestors = 324915398,324915287)